### PR TITLE
REST API compatiblity documentation

### PR DESCRIPTION
**Pull request by jakelandis**
related: #51816

Preview

https://elasticsearch_83487.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/rest-api-compatibility.html

https://elasticsearch_83487.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/migrating-8.0.html#breaking_80_rest_api_changes

https://elasticsearch_83487.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/setup-upgrade.html

### CODE REVIEW

File: docs/reference/rest-api/rest-api-compatibility.asciidoc [hunk 1]
```diff
@@ -0,0 +1,72 @@
+[[rest-api-compatibility]]
```
Comment (jakelandis) [inline docs/reference/rest-api/rest-api-compatibility.asciidoc:1]: I could use some help figuring out how to link up a couple TODO's below. Also, I left the line lengths long here in anticipation of many edits. I will reduce the length prior to merge.

File: docs/reference/rest-api/rest-api-compatibility.asciidoc [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+[source, text]
+------------------------------------------------------------
+"application/vnd.elasticsearch+json;compatible-with=7"
+"application/vnd.elasticsearch+yaml;compatible-with=7"
+"application/vnd.elasticsearch+smile;compatible-with=7"
+"application/vnd.elasticsearch+cbor;compatible-with=7"
+------------------------------------------------------------
+
+The link:/guide[officially supported Elasticsearch clients] can enable REST API compatibility for all requests. TODO: link to https://www.elastic.co/guide/en/elasticsearch/client/index.html
+To enable REST API compatibility for all requests set the environment variable `ELASTIC_CLIENT_APIVERSIONING` to true.
```
Comment (debadair) [inline docs/reference/rest-api/rest-api-compatibility.asciidoc:59]:
```suggestion
The [link:/guide](https://www.elastic.co/guide/en/elasticsearch/client/index.html)[officially supported {es} clients] can enable REST API compatibility for all requests. 

To enable REST API compatibility for all requests received by {es} set the environment variable `ELASTIC_CLIENT_APIVERSIONING` to true.
```

File: docs/reference/rest-api/rest-api-compatibility.asciidoc [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+=== REST API compatibility workflow
+
+The recommended workflow for upgrading to a new major and leverage REST API compatibility is as follows (from 7.x to 8.x):
+
+1. Upgrade any officially supported Elasticsearch clients to the latest 7.x version and enable REST API compatibility. TODO: link to https://www.elastic.co/guide/en/elasticsearch/client/index.html
+2. Use the {kibana-ref}/upgrade-assistant.html[Kibana upgrade assistant] and review all critical issues. Some critical issues may be mitigated by REST API compatibility.
+3. From the {kibana-ref}/upgrade-assistant.html[Kibana upgrade assistant] explore the deprecation logs and resolve all critical issues
+4. Upgrade Elasticsearch to 8.x
+5. Review the deprecation logs. If compatibility is invoked it will be logged to the deprecation logs with the category "compatible_api". Review the workflow associated with any invoked compatibility.
+6. Upgrade the official {es} clients to the matching 8.x version resolving any cases where compatibility needs to be manually resolved.
```
Comment (debadair) [inline docs/reference/rest-api/rest-api-compatibility.asciidoc:71]:
```suggestion

To leverage REST API compatibility during an upgrade from 7.17 to {version}:

1. Upgrade your https://www.elastic.co/guide/en/elasticsearch/client/index.html[{es} clients] to the latest 7.x version and enable REST API compatibility. 
2. Use the {kibana-ref}/upgrade-assistant.html[Upgrade Assistant] to review all critical issues and explore the deprecation logs. Some critical issues might be mitigated by REST API compatibility.
3. Resolve all critical issues before proceeding with the upgrade.
4. Upgrade Elasticsearch to {version}.
5. Review the deprecation logs for entries with the category `compatible_api1`. Review the workflow associated with the requests that relied on compatibility mode.
6. Upgrade your {es} clients to 8.x and resolve compatibility issues manually where needed. 
```

File: docs/reference/upgrade.asciidoc [hunk 4]
```diff
@@ -5,12 +5,12 @@
 --
 ifeval::["{release-state}"!="released"]
 [[upgrade-pre-release]]
-IMPORTANT: This documentation is for a pre-release of {es} {minor-version}. 
+IMPORTANT: This documentation is for a pre-release of {es} {minor-version}.
```
Comment (debadair) [inline docs/reference/upgrade.asciidoc:8]: I've got changes to this topic in flight. Revert these & I'll add the compatibility info where needed.  (And link to the new topic.)

**Review by debadair — CHANGES_REQUESTED** [review]
I added some editorial suggestions & tried to streamline this a bit. 

Let me know if you have questions!

File: docs/reference/upgrade.asciidoc (hunk 4 above)
Comment (jakelandis) [reply docs/reference/upgrade.asciidoc]: The whitespace changes were accidental and have been removed from the PR. Did you mean just the white space or the entire file ?

**Comment by jakelandis** [comment]
@elasticmachine run elasticsearch-ci/docs

File: docs/reference/rest-api/rest-api-compatibility.asciidoc [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+only required when a body is sent with the request. The following values are
+valid when communicating with a 7.x or 8.x {es} server:
+[source, text]
+------------------------------------------------------------
+"application/vnd.elasticsearch+json;compatible-with=7"
+"application/vnd.elasticsearch+yaml;compatible-with=7"
+"application/vnd.elasticsearch+smile;compatible-with=7"
+"application/vnd.elasticsearch+cbor;compatible-with=7"
+------------------------------------------------------------
+The [link:/guide](https://www.elastic.co/guide/en/elasticsearch/client/index.html)[officially supported {es} clients]
```
Comment (jakelandis) [inline docs/reference/rest-api/rest-api-compatibility.asciidoc:75]: @debadair - can you help out again with this link the build still claims this as broken:

```
13:02:16 INFO:build_docs:  /tmp/docsbuild/target_repo/html/en/elasticsearch/reference/master/rest-api-compatibility.html contains broken links to:
13:02:16 INFO:build_docs:   - en/elasticsearch/client/index.html)
```

Comment (debadair) [inline docs/reference/rest-api/rest-api-compatibility.asciidoc:75]:
```suggestion
The https://www.elastic.co/guide/en/elasticsearch/client/index.html[officially supported {es} clients]
```

**Review by debadair — APPROVED** [review]
Left a suggestion that should fix the link.

**Comment by jakelandis** [comment]
Thank you @debadair - your edits were really good !

**Review by pgomulka — COMMENTED** [review]
LGTM, sorry for the delay

**Pull request merged**